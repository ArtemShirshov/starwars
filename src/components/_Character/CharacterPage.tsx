import React from 'react';
import {
  Box,
  Button,
  Grid,
  Paper,
  Skeleton,
  Stack,
  TextField,
  Typography,
} from '@mui/material';

import { CharacterPageViewModel } from './CharacterPage.viewmodel';
import { Breadcrumbs } from './Breadcrumbs/Breadcrumbs';
import { Description } from './Description/Description';
import { descriptionItems } from './CharacterPage.constatns';

export const CharacterPage = () => {
  const {
    character,
    editing,
    id,
    loading,
    onChange,
    onReset,
    onSubmit,
    setEditing,
  } = CharacterPageViewModel();
  return (
    <>
      <Box my={2}>
        <Breadcrumbs loading={loading} page={character?.name} />
      </Box>

      <Paper
        sx={{
          mb: 10,
          overflow: 'hidden',
        }}
      >
        <Grid container>
          <Grid
            height={550}
            item
            sx={{
              minWidth: 400,
            }}
            xs="auto"
          >
            <img alt="" src={`/assets/characters/${id}.jpg`} />
          </Grid>
          <Grid
            item
            sx={{
              p: 2,
            }}
            xs
          >
            <Box mb={3}>
              {editing ? (
                <TextField
                  defaultValue={character?.name}
                  fullWidth
                  label="Name"
                  name="name"
                  onChange={onChange}
                />
              ) : (
                <Typography color="primary" variant="h4">
                  {loading ? <Skeleton /> : character?.name}
                </Typography>
              )}
            </Box>

            <Stack spacing={1}>
              {descriptionItems.map((item) => (
                <>
                  {character[item.key] ? (
                    <>
                      {editing ? (
                        <TextField
                          defaultValue={character[item.key]}
                          key={item.label}
                          label={item.label}
                          name={item.key}
                          onChange={onChange}
                          type={item.type}
                        />
                      ) : (
                        <Description
                          key={item.label}
                          label={item.label}
                          loading={loading}
                          value={`${character[item.key]}${
                            item.additionalLabel ? item.additionalLabel : ''
                          }`}
                        />
                      )}
                    </>
                  ) : null}
                </>
              ))}
            </Stack>

            <Box my={2}>
              {editing ? (
                <>
                  <Button onClick={onSubmit}>Save</Button>
                  <Button onClick={onReset}>Reset</Button>
                </>
              ) : (
                <Button
                  disabled={loading}
                  onClick={() => setEditing(true)}
                  type="submit"
                >
                  Edit
                </Button>
              )}
            </Box>
          </Grid>
        </Grid>
      </Paper>
    </>
  );
};
