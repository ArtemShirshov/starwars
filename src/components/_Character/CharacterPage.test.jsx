import React from 'react';
import renderer from 'react-test-renderer';

import { CharacterPage } from './CharacterPage';

jest.mock('@mui/material', () => ({
  ...jest.requireActual('@mui/material'),
  Typography: 'Typography',
  Grid: 'Grid',
  Stack: 'Stack',
  Box: 'Box',
  Skeleton: 'Skeleton',
  Button: 'Button',
  TextField: 'TextField',
}));

jest.mock('./Breadcrumbs/Breadcrumbs', () => ({
  Breadcrumbs: 'Breadcrumbs',
}));

jest.mock('./Description/Description', () => ({
  Description: 'Description',
}));

let mockData;

jest.mock('./CharacterPage.viewmodel', () => ({
  CharacterPageViewModel: () => mockData,
}));

beforeEach(() => {
  mockData = {
    character: {
      name: 'Luke',
      birth_year: 'Y1231',
    },
    editing: false,
    id: '1',
    loading: false,
    onChange: jest.fn(),
    onReset: jest.fn(),
    onSubmit: jest.fn(),
    setEditing: jest.fn(),
  };
});

describe('CharacterPage', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<CharacterPage />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with loading', () => {
    mockData = {
      ...mockData,
      loading: true,
    };

    const tree = renderer.create(<CharacterPage />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with editing', () => {
    mockData = {
      ...mockData,
      loading: false,
      editing: true,
    };

    const tree = renderer.create(<CharacterPage />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
