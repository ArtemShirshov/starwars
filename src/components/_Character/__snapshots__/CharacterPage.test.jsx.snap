// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterPage renders correctly 1`] = `
Array [
  <Box
    my={2}
  >
    <Breadcrumbs
      loading={false}
      page="Luke"
    />
  </Box>,
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-13it7l2-MuiPaper-root"
  >
    <Grid
      container={true}
    >
      <Grid
        height={550}
        item={true}
        sx={
          Object {
            "minWidth": 400,
          }
        }
        xs="auto"
      >
        <img
          alt=""
          src="/assets/characters/1.jpg"
        />
      </Grid>
      <Grid
        item={true}
        sx={
          Object {
            "p": 2,
          }
        }
        xs={true}
      >
        <Box
          mb={3}
        >
          <Typography
            color="primary"
            variant="h4"
          >
            Luke
          </Typography>
        </Box>
        <Stack
          spacing={1}
        >
          <Description
            label="Birth Year"
            loading={false}
            value="Y1231"
          />
        </Stack>
        <Box
          my={2}
        >
          <Button
            disabled={false}
            onClick={[Function]}
            type="submit"
          >
            Edit
          </Button>
        </Box>
      </Grid>
    </Grid>
  </div>,
]
`;

exports[`CharacterPage renders correctly with editing 1`] = `
Array [
  <Box
    my={2}
  >
    <Breadcrumbs
      loading={false}
      page="Luke"
    />
  </Box>,
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-13it7l2-MuiPaper-root"
  >
    <Grid
      container={true}
    >
      <Grid
        height={550}
        item={true}
        sx={
          Object {
            "minWidth": 400,
          }
        }
        xs="auto"
      >
        <img
          alt=""
          src="/assets/characters/1.jpg"
        />
      </Grid>
      <Grid
        item={true}
        sx={
          Object {
            "p": 2,
          }
        }
        xs={true}
      >
        <Box
          mb={3}
        >
          <TextField
            defaultValue="Luke"
            fullWidth={true}
            label="Name"
            name="name"
            onChange={[MockFunction]}
          />
        </Box>
        <Stack
          spacing={1}
        >
          <TextField
            defaultValue="Y1231"
            label="Birth Year"
            name="birth_year"
            onChange={[MockFunction]}
          />
        </Stack>
        <Box
          my={2}
        >
          <Button
            onClick={[MockFunction]}
          >
            Save
          </Button>
          <Button
            onClick={[MockFunction]}
          >
            Reset
          </Button>
        </Box>
      </Grid>
    </Grid>
  </div>,
]
`;

exports[`CharacterPage renders correctly with loading 1`] = `
Array [
  <Box
    my={2}
  >
    <Breadcrumbs
      loading={true}
      page="Luke"
    />
  </Box>,
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-13it7l2-MuiPaper-root"
  >
    <Grid
      container={true}
    >
      <Grid
        height={550}
        item={true}
        sx={
          Object {
            "minWidth": 400,
          }
        }
        xs="auto"
      >
        <img
          alt=""
          src="/assets/characters/1.jpg"
        />
      </Grid>
      <Grid
        item={true}
        sx={
          Object {
            "p": 2,
          }
        }
        xs={true}
      >
        <Box
          mb={3}
        >
          <Typography
            color="primary"
            variant="h4"
          >
            <Skeleton />
          </Typography>
        </Box>
        <Stack
          spacing={1}
        >
          <Description
            label="Birth Year"
            loading={true}
            value="Y1231"
          />
        </Stack>
        <Box
          my={2}
        >
          <Button
            disabled={true}
            onClick={[Function]}
            type="submit"
          >
            Edit
          </Button>
        </Box>
      </Grid>
    </Grid>
  </div>,
]
`;
